src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -475,22 +475,45 @@ void ObjectSynchronizer::jni_exit(oop obj, TRAPS) {
 // -----------------------------------------------------------------------------
 // Internal VM locks on java objects
 // standard constructor, allows locking failures
-ObjectLocker::ObjectLocker(Handle obj, JavaThread* thread) : _npm(thread) {
-  _thread = thread;
+ObjectLocker::ObjectLocker(Handle obj, TRAPS) : _thread(THREAD), _obj(obj),
+  _npm(_thread, _thread->at_preemptable_init() /* ignore_mark */), _skip_exit(false) {
+  assert(!_thread->preempting(), "");
+
   _thread->check_for_valid_safepoint_state();
-  _obj = obj;
 
   if (_obj() != nullptr) {
     ObjectSynchronizer::enter(_obj, &_lock, _thread);
+
+    if (_thread->preempting()) {
+      // If preemption was cancelled we acquired the monitor after freezing
+      // the frames. Redoing the vm call later in thaw will require us to
```

Comment:
I don't quite follow the control flow here - at what point have we frozen any frames? Was that deep inside `enter`?